x-pack/plugins/data_enhanced/common/search/poll_search.ts

```diff
@@ -13,19 +13,28 @@ import type { IAsyncSearchOptions } from './types';
 
 export const pollSearch = <Response extends IKibanaSearchResponse>(
   search: () => Promise<Response>,
+  cancel?: () => void,
   { pollInterval = 1000, ...options }: IAsyncSearchOptions = {}
 ): Observable<Response> => {
-  const aborted = options?.abortSignal
-    ? abortSignalToPromise(options?.abortSignal)
-    : { promise: NEVER, cleanup: () => {} };
+  return defer(() => {
+    const aborted = options?.abortSignal
+      ? abortSignalToPromise(options?.abortSignal)
+      : { promise: new Promise(() => {}), cleanup: () => {} };
 
-  return from(search()).pipe(
-    expand(() => timer(pollInterval).pipe(switchMap(search))),
-    tap((response) => {
-      if (isErrorResponse(response)) throw new AbortError();
-    }),
-    takeWhile<Response>(isPartialResponse, true),
-    takeUntil<Response>(from(aborted.promise)),
-    finalize(aborted.cleanup)
-  );
+    aborted.promise.catch(() => {
+      if (cancel) cancel();
+    });
+
+    return from(search()).pipe(
+      expand(() => timer(pollInterval).pipe(switchMap(search))),
+      tap((response) => {
+        if (isErrorResponse(response)) {
+          throw response ? new Error('Received partial response') : new AbortError();
+        }
+      }),
+      takeWhile<Response>(isPartialResponse, true),
+      takeUntil<Response>(from(aborted.promise)),
+      finalize(aborted.cleanup)
+    );
```

Comment:
Will this result in a memory leak since it is never removing the event listener from the `abortSignal`?

Reply:
I would suspect not, because nothing holds a reference to the abort signal once the observable goes out of scope. The gc should be able to clean it up afterwards.